xlive/Blam/Engine/saved_games/saved_game_files_async_windows.cpp

```diff
@@ -0,0 +1,252 @@
+t_saved_games_async_helper_read_file_internal p_saved_games_async_helper_read_file_internal;
+
+typedef void(__cdecl* t_saved_games_async_helper_read_file_callback)(int32 a1, int8* context, int32 data_size);
+t_saved_games_async_helper_read_file_callback p_saved_games_async_helper_read_file_callback;
+
+typedef void(__cdecl* t_saved_games_async_helper_read_file)(uint32 enumerated_index, int8* buffer, uint32 buffer_size);
+t_saved_games_async_helper_read_file p_saved_games_async_helper_read_file;
+
+/* private code */
+
+bool __cdecl saved_games_async_helper_write_file(int32 enumerated_index, wchar_t* new_display_name, void* buffer, uint32 buffer_size, int8* completion)
+{
+	return p_saved_games_async_helper_write_file(enumerated_index, new_display_name, buffer, buffer_size, completion);
+}
+
+bool __cdecl saved_games_async_helper_async_read_create_task(void* unk, void* buffer, uint32 buffer_size, bool unk_flag, int8* in_out_completion)
+{
+	return INVOKE(0x9b149, 0, saved_games_async_helper_async_read_create_task, unk, buffer, buffer_size, unk_flag, in_out_completion);
+}
+
+bool __cdecl saved_games_async_helper_read_file_internal(int enumerated_index, void* buffer, unsigned int buffer_size, int8* in_out_completion)
+{
+		s_saved_game_main_menu_globals* saved_game_globals = saved_game_main_menu_globals_get();
+
+		auto abs_index = (enumerated_index >> 8) & 0x1FFF;
+		auto last_index = saved_game_globals->default_save_files.get_count() - 1;
+
+		// 0x200000 in the index flags that the index is a default file
+		if ((enumerated_index & 0x200000) != 0 && saved_game_globals)
+		{
+
+			// Removed check of buffer_size so it can read the entry from the globals
+			// potentially unsafe will have to come up with a way to verify if this was actually needed or not.
+			if ((abs_index <= last_index || abs_index == last_index))
+			{
+				auto test = saved_game_globals->default_save_files[abs_index];
+				csmemcpy(buffer, saved_game_globals->default_save_files[abs_index]->buffer, buffer_size);
+				in_out_completion[2] = 1;
+				*(in_out_completion + 1) = 0;
+				*in_out_completion = 1;
+				*(in_out_completion + 514) = 0x3F800000;
+				return true;
+			}
+		}
+		else
+		{
+			s_saved_game_main_menu_globals_save_file_info file_info {};
+			wchar_t file_path[256];
+			char flattened_file_path[256];
+			if(saved_games_get_file_info(&file_info, enumerated_index))
+			{
+				if(file_info.unk_5 != NONE)
+				{
+					if(saved_games_append_file_type_to_path(file_info.file_path, file_info.type, file_path))
+					{
+						WideCharToMultiByte(0xFDE9u, 0, file_path, -1, flattened_file_path, 256, 0, 0);
+						bool result =  saved_games_async_helper_async_read_create_task(flattened_file_path, buffer, buffer_size, 0, in_out_completion);
+
+						return result;
+					}
+				}
+			}
+		}
+}
+
+void saved_games_async_helper_read_file_success(int8* a1, s_file_reference* file_reference, int8* buffer, uint32 buffer_size)
+{
+	int32* unk_1 = ((int32*)a1 + 1);
```

Comment:
implement the structure for this

Reply:
it's an async related structure, i'll add a comment to replace with the structure when async is done.